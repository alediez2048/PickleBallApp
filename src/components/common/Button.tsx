import React from "react";
import {
  TouchableOpacity,
  StyleSheet,
  View,
  ActivityIndicator,
  TouchableOpacityProps,
  Platform,
} from "react-native";
import { ThemedText } from "@/components/common/ThemedText";

interface ButtonProps extends TouchableOpacityProps {
  variant?: "primary" | "secondary" | "outline" | "danger";
  size?: "small" | "medium" | "large";
  loading?: boolean;
  fullWidth?: boolean;
  children: React.ReactNode;
}

export const Button: React.FC<ButtonProps> = ({
  variant = "primary",
  size = "medium",
  loading = false,
  disabled = false,
  fullWidth = false,
  style,
  children,
  accessibilityLabel,
  accessibilityHint,
  onPress,
  ...props
}) => {
  const isDisabled = disabled || loading;

  const getButtonStyle = () => {
    const variantStyles = {
      primary: styles.primary,
      secondary: styles.secondary,
      outline: styles.outline,
      danger: styles.danger,
    };

    const sizeStyles = {
      small: styles.small,
      medium: styles.medium,
      large: styles.large,
    };

    const computedStyles = [
      styles.base,
      variantStyles[variant],
      sizeStyles[size],
      fullWidth && styles.fullWidth,
      isDisabled && styles.disabled,
      style,
    ];

    // Debug logging
    console.debug("[Button Render]", {
      text: typeof children === "string" ? children : "Custom children",
      variant,
      size,
      isDisabled,
      fullWidth,
      customStyle: style,
      computedStyles: computedStyles
        .filter(
          (s): s is NonNullable<typeof s> => s !== null && s !== undefined
        )
        .map((s) => (typeof s === "object" ? Object.keys(s) : [])),
    });

    return computedStyles;
  };

  const getTextStyle = () => {
    const variantTextStyles = {
      primary: styles.primaryText,
      secondary: styles.secondaryText,
      outline: styles.outlineText,
      danger: styles.dangerText,
    };

    const computedTextStyles = [
      styles.text,
      variantTextStyles[variant],
      isDisabled && styles.disabledText,
    ];

    return computedTextStyles;
  };

  const buttonLabel =
    typeof children === "string" ? children : accessibilityLabel;

  return (
    <TouchableOpacity
      testID='button'
      style={getButtonStyle()}
      disabled={isDisabled}
      onPress={onPress}
      accessibilityRole='button'
      accessibilityLabel={buttonLabel}
      accessibilityHint={accessibilityHint}
      accessibilityState={{
        disabled: isDisabled,
        busy: loading,
      }}
      {...props}
    >
      {loading ? (
        <View style={styles.loadingContainer}>
          <ActivityIndicator
            testID='loading-spinner'
            color={variant === "outline" ? "#4CAF50" : "#FFFFFF"}
            accessibilityLabel='Loading'
          />
        </View>
      ) : typeof children === "string" ? (
        <ThemedText style={getTextStyle()}>{children}</ThemedText>
      ) : (
        children
      )}
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  base: {
    borderRadius: 25,
    alignItems: "center",
    justifyContent: "center",
    flexDirection: "row",
    borderWidth: 1,
    borderColor: "transparent",
  },
  primary: {
    backgroundColor: "#4CAF50",
    borderColor: "#4CAF50",
  },
  secondary: {
    backgroundColor: "#2196F3",
    borderColor: "#2196F3",
  },
  outline: {
    backgroundColor: "transparent",
    borderColor: "#4CAF50",
  },
  danger: {
    backgroundColor: "#DC2626",
    borderColor: "#DC2626",
  },
  small: {
    paddingVertical: 8,
    paddingHorizontal: 20,
    minHeight: 36,
  },
  medium: {
    paddingVertical: 12,
    paddingHorizontal: 28,
    minHeight: 44,
  },
  large: {
    paddingVertical: 14,
    paddingHorizontal: 32,
    minHeight: 52,
  },
  fullWidth: {
    width: "100%",
  },
  disabled: {
    opacity: 0.5,
    backgroundColor: "#E0E0E0",
    borderColor: "#E0E0E0",
  },
  text: {
    fontSize: 16,
    fontWeight: "600",
    textAlign: "center",
    letterSpacing: 0.5,
  },
  primaryText: {
    color: "#FFFFFF",
  },
  secondaryText: {
    color: "#FFFFFF",
  },
  outlineText: {
    color: "#4CAF50",
  },
  dangerText: {
    color: "#FFFFFF",
  },
  disabledText: {
    color: "#666666",
  },
  loadingContainer: {
    minHeight: 24,
    justifyContent: "center",
    alignItems: "center",
  },
});
