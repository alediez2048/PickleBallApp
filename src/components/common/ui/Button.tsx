import React from 'react';
import { TouchableOpacity, Text, TouchableOpacityProps, ActivityIndicator, View, StyleSheet } from 'react-native';
import { withMemo } from '@/components/hoc/withMemo';

interface ButtonProps extends TouchableOpacityProps {
  variant?: 'primary' | 'secondary';
  size?: 'sm' | 'md' | 'lg';
  children: React.ReactNode;
  loading?: boolean;
  accessibilityLabel?: string;
  accessibilityHint?: string;
}

const ButtonComponent: React.FC<ButtonProps> = ({
  variant = 'primary',
  size = 'md',
  children,
  disabled,
  loading,
  style,
  accessibilityLabel,
  accessibilityHint,
  ...props
}) => {
  const buttonStyles = [
    styles.button,
    styles[`${size}Button`],
    variant === 'primary' ? styles.primaryButton : styles.secondaryButton,
    disabled && styles.disabledButton,
    style,
  ];

  const textStyles = [
    styles.text,
    styles[`${size}Text`],
    variant === 'primary' ? styles.primaryText : styles.secondaryText,
    disabled && styles.disabledText,
  ];

  // Determine the button's label for screen readers
  const buttonLabel = accessibilityLabel || (typeof children === 'string' ? children : undefined);

  return (
    <TouchableOpacity
      style={buttonStyles}
      disabled={disabled || loading}
      accessibilityRole="button"
      accessibilityLabel={buttonLabel}
      accessibilityHint={accessibilityHint}
      accessibilityState={{
        disabled: disabled || loading,
        busy: loading,
      }}
      {...props}
    >
      {loading ? (
        <View style={styles.loadingContainer}>
          <ActivityIndicator 
            color={variant === 'primary' ? '#fff' : '#000'} 
            accessibilityLabel="Loading"
          />
        </View>
      ) : (
        typeof children === 'string' ? (
          <Text 
            style={textStyles}
            numberOfLines={1}
            accessibilityRole="text"
          >
            {children}
          </Text>
        ) : (
          children
        )
      )}
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  button: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 8,
    paddingHorizontal: 16,
    marginVertical: 6,
  },
  smButton: {
    height: 32,
    paddingHorizontal: 12,
  },
  mdButton: {
    height: 40,
    paddingHorizontal: 16,
  },
  lgButton: {
    height: 48,
    paddingHorizontal: 20,
  },
  primaryButton: {
    backgroundColor: '#000',
  },
  secondaryButton: {
    backgroundColor: '#fff',
    borderWidth: 1,
    borderColor: '#E5E7EB',
  },
  disabledButton: {
    opacity: 0.5,
  },
  text: {
    fontWeight: '500',
    textAlign: 'center',
  },
  smText: {
    fontSize: 14,
  },
  mdText: {
    fontSize: 16,
  },
  lgText: {
    fontSize: 18,
  },
  primaryText: {
    color: '#fff',
  },
  secondaryText: {
    color: '#000',
  },
  disabledText: {
    opacity: 0.7,
  },
  loadingContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
  },
});

export const Button = withMemo(ButtonComponent); 