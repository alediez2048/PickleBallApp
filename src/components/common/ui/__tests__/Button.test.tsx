import React from 'react';
import renderer from 'react-test-renderer';
import { Button } from '../Button';
import { TouchableOpacity } from 'react-native';

// Mock UIProvider context
jest.mock('@/contexts/UIContext', () => ({
  UIProvider: ({ children }: { children: React.ReactNode }) => children,
  useUIState: () => ({
    colorScheme: 'light',
    theme: 'default'
  }),
  useThemedColor: () => ({
    primary: '#4CAF50',
    secondary: '#2196F3',
    text: '#333333',
    background: '#FFFFFF'
  })
}));

describe('Button', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with default props', () => {
    const tree = renderer
      .create(<Button>Test Button</Button>)
      .toJSON();
    
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with different sizes', () => {
    const smallTree = renderer
      .create(<Button size="small">Small Button</Button>)
      .toJSON();
    
    const mediumTree = renderer
      .create(<Button size="medium">Medium Button</Button>)
      .toJSON();
    
    const largeTree = renderer
      .create(<Button size="large">Large Button</Button>)
      .toJSON();
    
    expect(smallTree).toMatchSnapshot('small size');
    expect(mediumTree).toMatchSnapshot('medium size');
    expect(largeTree).toMatchSnapshot('large size');
  });

  it('renders correctly with different variants', () => {
    const primaryTree = renderer
      .create(<Button variant="primary">Primary Button</Button>)
      .toJSON();
    
    const secondaryTree = renderer
      .create(<Button variant="secondary">Secondary Button</Button>)
      .toJSON();
    
    const outlineTree = renderer
      .create(<Button variant="outline">Outline Button</Button>)
      .toJSON();
    
    expect(primaryTree).toMatchSnapshot('primary variant');
    expect(secondaryTree).toMatchSnapshot('secondary variant');
    expect(outlineTree).toMatchSnapshot('outline variant');
  });

  it('renders correctly in loading state', () => {
    const tree = renderer
      .create(<Button loading>Loading Button</Button>)
      .toJSON();
    
    expect(tree).toMatchSnapshot('loading state');
  });

  it('renders correctly in disabled state', () => {
    const tree = renderer
      .create(<Button disabled>Disabled Button</Button>)
      .toJSON();
    
    expect(tree).toMatchSnapshot('disabled state');
  });

  it('renders correctly with full width', () => {
    const tree = renderer
      .create(<Button fullWidth>Full Width Button</Button>)
      .toJSON();
    
    expect(tree).toMatchSnapshot('full width');
  });

  it('renders correctly with accessibility props', () => {
    const tree = renderer
      .create(
        <Button 
          accessibilityLabel="Custom Label"
          accessibilityHint="This button does something"
        >
          Accessible Button
        </Button>
      )
      .toJSON();
    
    expect(tree).toMatchSnapshot('accessibility props');
  });

  it('renders correctly with icon children', () => {
    const tree = renderer
      .create(
        <Button accessibilityLabel="Icon Button">
          <React.Fragment>Icon</React.Fragment>
        </Button>
      )
      .toJSON();
    
    expect(tree).toMatchSnapshot('icon children');
  });

  it('renders correctly in loading and disabled state', () => {
    const tree = renderer
      .create(<Button loading disabled>Loading Disabled Button</Button>)
      .toJSON();
    
    expect(tree).toMatchSnapshot('loading and disabled');
  });
}); 