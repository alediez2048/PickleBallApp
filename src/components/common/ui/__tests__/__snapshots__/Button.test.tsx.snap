// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly in disabled state: disabled state 1`] = `null`;

exports[`Button renders correctly in loading and disabled state: loading and disabled 1`] = `null`;

exports[`Button renders correctly in loading state: loading state 1`] = `null`;

exports[`Button renders correctly with accessibility props: accessibility props 1`] = `null`;

exports[`Button renders correctly with default props 1`] = `null`;

exports[`Button renders correctly with different sizes: large size 1`] = `null`;

exports[`Button renders correctly with different sizes: medium size 1`] = `null`;

exports[`Button renders correctly with different sizes: small size 1`] = `null`;

exports[`Button renders correctly with different variants: outline variant 1`] = `null`;

exports[`Button renders correctly with different variants: primary variant 1`] = `null`;

exports[`Button renders correctly with different variants: secondary variant 1`] = `null`;

exports[`Button renders correctly with full width: full width 1`] = `null`;

exports[`Button renders correctly with icon children: icon children 1`] = `null`;
